### DISCUSSION TITLE
Cygwin and Mingw (Windows) support

### DISCUSSION THREAD

**Discussion by rasa** [przemoc/metastore#56] [Ideas]
I have it compiling and working in Cygwin and Mingw in
https://github.com/rasa/metastore.git
but since Windows doesn't really support [`uid`](https://github.com/rasa/metastore/blob/master/src/metaentry.c#L226)'s or [`gid`](https://github.com/rasa/metastore/blob/master/src/metaentry.c#L235)'s, we need to tweak the code to use [`SID`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379594%28v=vs.85%29.aspx)'s instead (at least for MingW, Cygwin works around this somehow).

Cygwin's SID > uid mapping logic is at:
https://cygwin.com/cygwin-ug-net/ntsec.html

**Comment by przemoc** [comment]
Thank you for your effort in bringing metastore to cygwin, but OS-dependent stuff has to be though-out for many systems before adding even one. I don't want bandaids and more `#ifdef` spaghetti.

As I wrote in [comment](https://github.com/przemoc/metastore/issues/39#issuecomment-170145951) to issue #39, I want to ultimately make metastore more modular. There should be some generic POSIX module (e.g. for ownership, perms), and OS-specific ones (e.g. for xattr). Cygwin should be possibly treated as POSIX alternative (as it's kind of mixed world), similarly native Windows-dedicated module if it will ever be created (but it wouldn't use uid/gid nomenclature there, etc.). Sorry that it's not there yet, but I'm not focusing on metastore lately.

Because of what I wrote above, I have to reject your PR atm. It wouldn't be merged as-is also for other reasons. To name a few:
Various changes in one commit are bad, so please try to put only one logical change in each commit. Embedding files from other project "along the way", i.e. during introducing other changes, is also bad, so please always perform such actions in commits solely dedicated to them. In future please also develop new stuff for PR in non-master branch, ideally named appropriately.

Why you even embedded strmode.h, strmode.c in the first place? (If it was done in its own commit, you could (and actually should) try to explain it in commit message.)

**Comment by zeluspudding** [comment]
Would love a metastore like tool for Windows. Can you suggest other tools that store meta info w/ git that works on windows?

**Comment by rasa** [comment]
I attached my Windows versions (Mingw and Cygwin) of metastore to https://github.com/rasa/metastore/releases/tag/95d4e2d
